### DISCUSSION TITLE
What is the difference between a route layer and a layer?

### DISCUSSION THREAD

**Discussion by Lydanne** [tokio-rs/axum#2878] [General]
I did not find the relevant documentation.

**Comment by dayvejones** [comment]
Hey!


[layer](https://docs.rs/axum/latest/axum/routing/struct.Router.html#method.layer) - Apply a [tower::Layer](https://docs.rs/tower-layer/0.3.2/x86_64-unknown-linux-gnu/tower_layer/trait.Layer.html) to all routes in the router

[route_layer](https://docs.rs/axum/latest/axum/routing/struct.Router.html#method.route_layer) - Apply a [tower::Layer](https://docs.rs/tower-layer/0.3.2/x86_64-unknown-linux-gnu/tower_layer/trait.Layer.html) to the router that will only run if the request matches a route.


For example:

```
let app = Router::new()
    .route("/foo", get(|| async {}))
    .route_layer(ValidateRequestHeaderLayer::bearer("password"));

// `GET /foo` with a valid token will receive `200 OK`
// `GET /foo` with a invalid token will receive `401 Unauthorized`
// `GET /not-found` with a invalid token will receive `404 Not Found`
```

The route "/not-found" is missing from the router. The **layer** will process it, but the **route_layer** will not.

**Reply by Lydanne** [reply]
Thanks for your reply, I understand❤️

**Reply by sumit03guha** [reply]
I am still a bit confused. What is the difference between the following two pieces of code snippet?
```
let auth_router = Router::new()
        .route("/signup", post(signup))
        .route("/signin", post(signin))
        .route(
            "/protected",
            get(protected).route_layer(from_fn(login_required)),
        );
```
and 
```
let auth_router = Router::new()
        .route("/signup", post(signup))
        .route("/signin", post(signin))
        .route(
            "/protected",
            get(protected).layer(from_fn(login_required)),
        );
```

@dayvejones help here please..

**Reply by jplatte** [reply]
In the first snippet, the middleware only runs for `GET /protected`. In the second snippet, the middleware runs for `GET /protected`, but also `POST /protected`, `PUT /protected` and so on.
